### PR TITLE
Code Improvement: Error handling on sam cli calls

### PR DESCRIPTION
**Pull request by awschristou**
## Types of changes

- [x] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)

## Description

* Consolidated all of the SAM CLI error handling
* Fixed problem with `TestLogger` that didn't properly clean up or read from the log file

## Testing

* Ran tests
* Ran toolkit
  * Created new sam app
  * local run/debug of sam app
  * deploy sam app

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/loggerUtils.ts [hunk 1]
```diff
@@ -11,39 +11,38 @@ import * as filesystemUtilities from './filesystemUtilities'
...
+            await del(this.logFolder, { force: true })
         }
     }
 
     public async logContainsText(str: string): Promise<boolean> {
-        const logText = await filesystemUtilities.readFileAsString(this._logfile as string)
+        const logText = await filesystemUtilities.readFileAsString(this.logfilePath)
 
-        return(logText.includes(str))
+        return (logText.includes(str))
```
Comment (bryceitoc9) [inline src/shared/loggerUtils.ts]: Was about to ask why there were parenthesis around the return statement and then I realized that I was the one that put those in :/

Mind fixing that for me?

File: src/shared/sam/cli/samCliInvokerUtils.ts [hunk 2]
```diff
@@ -6,9 +6,37 @@
...
 }
+
+export function logAndThrowIfUnexpectedExitCode(
+    processResult: ChildProcessResult,
+    expectedExitCode: number = 0,
+    logger: BasicLogger = getLogger(),
+): void {
+    if (processResult.exitCode === expectedExitCode) { return }
+
+    logger.error(`Unexpected exitcode (${processResult.exitCode})`)
```
Comment (bryceitoc9) [inline src/shared/sam/cli/samCliInvokerUtils.ts]: `Unexpected exitcode (${processResult.exitCode}), expecting (${expectedExitCode})` ?

File: src/shared/sam/cli/samCliInit.ts [hunk 3]
```diff
@@ -20,25 +18,12 @@ export async function runSamCliInit(
...
-    console.error(`Exit code: ${exitCode}`)
-    console.error(`Error: ${error}`)
-    console.error(`stderr: ${stderr}`)
-    console.error(`stdout: ${stdout}`)
-
-    const logger: Logger = getLogger()
-    const err = new Error(`sam init encountered an error: ${error && error.message || stderr || stdout}`)
-    logger.error(err)
-    throw err
+    logAndThrowIfUnexpectedExitCode(childProcessResult, 0)
```
Comment (bryceitoc9) [inline src/shared/sam/cli/samCliInit.ts]: I feel like you should leave out the `, 0`. It feels like that should be understood.

At what points would we expect something other than 0?

File: src/shared/sam/cli/samCliBuild.ts [hunk 4]
```diff
@@ -73,6 +76,7 @@ export class SamCliBuildInvocation {
             skipPullImage = false,
             ...params
         }: SamCliBuildInvocationArguments,
+        private readonly file: FileFunctions = getDefaultFileFunctions()
```
Comment (mpiroc) [inline src/shared/sam/cli/samCliBuild.ts]: Nit: For consistency, please wrap this in a context, i.e. `private readonly context: { file: FileFunctions } = ...`

File: src/shared/sam/cli/samCliInit.ts (hunk 3 above)
Comment (awschristou) [reply src/shared/sam/cli/samCliInit.ts]: Best to be explicit and make no assumptions about what default `logAndThrowIfUnexpectedExitCode` contains.

Comment (bryceitoc9) [reply src/shared/sam/cli/samCliInit.ts]: Thoughts on something like:
```
export function logAndThrowIfError(childProcessResult) {
    logAndThrowIfUnexpectedExitCode(childProcessResult, 0)
}
```

It just doesn't feel like there's much reason to watch for anything aside from 0.

File: src/shared/loggerUtils.ts [hunk 5]
```diff
@@ -11,39 +11,38 @@ import * as filesystemUtilities from './filesystemUtilities'
...
-
     // initializes a default logger. This persists through all tests.
     // initializing a default logger means that any tested files with logger statements will work.
     // as a best practice, please initialize a TestLogger before running tests on a file with logger statements.
-    private constructor(logfile: string, logger: l.Logger) {
-        this._logger = logger
-        this._logfile = logfile
-    }
+    private constructor(
+        private readonly logFolder: string,
```
[the remark this answers is not in the document]
Comment (awschristou) [reply src/shared/loggerUtils.ts]: This class is not well written to handle cases other than the one hardcoded into `createTestLogger`. I'll see if I can clean it up a little.

**Pull request merged**